File: VB/GridControlCRUD.Common/ViewModel - VB/AsyncCollectionViewModel.vb
Imports DevExpress.CRUD.DataModel
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Mvvm.Xpf
Imports System.Collections.Generic
Imports System.Threading.Tasks

Namespace DevExpress.CRUD.ViewModel
	Public MustInherit Class AsyncCollectionViewModel(Of T As Class)
		Inherits ViewModelBase

		Private ReadOnly dataProvider As IDataProvider(Of T)

		Protected Sub New(ByVal dataProvider As IDataProvider(Of T))
			Me.dataProvider = dataProvider
			StartRefresh()
		End Sub

		Public Property Entities() As IList(Of T)
			Get
				Return GetValue(Of IList(Of T))()
			End Get
			Private Set(ByVal value As IList(Of T))
				SetValue(value)
			End Set
		End Property
		Public Property EntitiesErrorMessage() As String
			Get
				Return GetValue(Of String)()
			End Get
			Private Set(ByVal value As String)
				SetValue(value)
			End Set
		End Property
		Public Property IsLoading() As Boolean
			Get
				Return GetValue(Of Boolean)()
			End Get
			Private Set(ByVal value As Boolean)
				SetValue(value)
			End Set
		End Property

		Private Async Sub StartRefresh()
			Await OnRefreshAsync()
		End Sub

		<Command>
		Public Sub OnRefresh(ByVal args As RefreshArgs)
			args.Result = OnRefreshAsync()
		End Sub
		Private Async Function OnRefreshAsync() As Task
			IsLoading = True
			Try
				Await Task.WhenAll(RefreshEntities(), OnRefreshCoreAsync())
			Finally
				IsLoading = False
			End Try
		End Function
		Private Async Function RefreshEntities() As Task
			Try
				Entities = Await dataProvider.ReadAsync()
				EntitiesErrorMessage = Nothing
			Catch
				Entities = Nothing
				EntitiesErrorMessage = "An error has occurred while establishing a connection to the database. Press F5 to retry the connection."
			End Try
		End Function
		Protected Overridable Function OnRefreshCoreAsync() As Task
			Return Task.CompletedTask
		End Function

		<Command>
		Public Sub OnUpdateRow(ByVal args As RowValidationArgs)
			args.ResultAsync = OnUpdateRowAsync(CType(args.Item, T), args.IsNewItem)
		End Sub
		Private Async Function OnUpdateRowAsync(ByVal entity As T, ByVal isNew As Boolean) As Task(Of ValidationErrorInfo)
			If isNew Then
				Await dataProvider.CreateAsync(entity)
			Else
				Await dataProvider.UpdateAsync(entity)
			End If
			Return Nothing
		End Function

		<Command>
		Public Sub OnDelete(ByVal args As RowDeleteArgs)
			dataProvider.Delete(DirectCast(args.Row, T))
		End Sub
	End Class
End Namespace

File: CS/GridControlCRUD.Common/ViewModel/AsyncCollectionViewModel.cs
﻿using DevExpress.CRUD.DataModel;
using DevExpress.Mvvm;
using DevExpress.Mvvm.DataAnnotations;
using DevExpress.Mvvm.Xpf;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace DevExpress.CRUD.ViewModel {
    public abstract class AsyncCollectionViewModel<T> : ViewModelBase where T : class {
        readonly IDataProvider<T> dataProvider;

        public IMessageBoxService MessageBoxService { get { return GetService<IMessageBoxService>(); } }

        protected AsyncCollectionViewModel(IDataProvider<T> dataProvider) {
            this.dataProvider = dataProvider;
            StartRefresh();
        }

        public IList<T> Entities {
            get => GetValue<IList<T>>();
            private set => SetValue(value);
        }
        public string EntitiesErrorMessage {
            get => GetValue<string>();
            private set => SetValue(value);
        }
        public bool IsLoading {
            get => GetValue<bool>();
            private set => SetValue(value);
        }

        async void StartRefresh() {
            await OnRefreshAsync();
        }

        [Command]
        public void OnRefresh(RefreshArgs args) {
            args.Result = OnRefreshAsync();
        }
        async Task OnRefreshAsync() {
            IsLoading = true;
            try {
                await Task.WhenAll(RefreshEntities(), OnRefreshCoreAsync());
            } finally {
                IsLoading = false;
            }
        }
        async Task RefreshEntities() {
            try {
                Entities = await dataProvider.ReadAsync();
                EntitiesErrorMessage = null;
            } catch {
                Entities = null;
                EntitiesErrorMessage = "An error has occurred while establishing a connection to the database. Press F5 to retry the connection.";
            }
        }
        protected virtual Task OnRefreshCoreAsync() {
            return Task.CompletedTask;
        }

        [Command]
        public void OnUpdateRow(RowValidationArgs args) {
            args.ResultAsync = OnUpdateRowAsync((T)args.Item, args.IsNewItem);
        }
        async Task<ValidationErrorInfo> OnUpdateRowAsync(T entity, bool isNew) {
            if(isNew)
                await dataProvider.CreateAsync(entity);
            else
                await dataProvider.UpdateAsync(entity);
            return null;
        }

        [Command]
        public void OnDeleteRow(DeleteRowsValidationArgs args) {
            var row = (T)args.Items[0];
            if(row == null)
                return;
            if(MessageBoxService.ShowMessage("Are you sure you want to delete this row?", "Delete Row", MessageButton.OKCancel) == MessageResult.Cancel) {
                args.Result = "Canceled";
            }
            dataProvider.Delete(row);
        }
    }
}
